### DISCUSSION TITLE
API

### DISCUSSION THREAD

**Discussion by thkruz** [thkruz/keeptrack.space#916] [Ideas]
I have had a couple of requests to create a public API. Is there anything specific people are looking for?

The most obvious would be the TLEs. I can see some people wanting to be able to pull the additional data I have on the satellites (owner/operator/rocket/etc). 

I need to work out how to create accounts and provide API keys, but definitely chime in if you have requests.

**Comment by awatson1978** [comment]
Have you ever considered supporting URL params?  So that users can open up specific scenes?  Bookmark how the viewport or scenario is loaded into the viewer?

**Reply by thkruz** [reply]
https://github.com/thkruz/keeptrack.space/blob/main/src%2Fengine%2Finput%2Furl-manager.ts

Quick AI generated cheat sheet on what it already does. If there is more you wanted please let me know. For the official/pro version I am also working on a scenario manager so that you can serialize the whole scene and export/import it as a single file. 

# URL Parameters: Import / Export Reference

`UrlManager` reads parameters on load (`parseGetVariables`) and writes them back as the scene changes (`updateURL`). This lets a URL capture and restore a viewer state.

## Round-trip (imported and exported)

| Param | Effect | Export condition |
| --- | --- | --- |
| `date` | Simulation time. Accepts ISO 8601 or a 13-digit unix timestamp in ms | When `isMaxData`, or when `staticOffset` is beyond +/-1000 |
| `rate` | Propagation rate, clamped to 0 to 1000 | Only when not roughly 1.0 (outside 0.99 to 1.01) |
| `sat` | Selects a satellite by SCC number | When a satellite is selected (does not work for VIMPEL objects) |
| `search` | Runs a search on load | When the search string is non-empty |
| `limitSats` | Catalog limit string | When set |
| `regime` | Orbital regime filter, comma separated | When the filter is non-empty |
| `pitch` / `yaw` | Camera angles in degrees | When non-zero. Uses FTS angles when locked to a satellite, otherwise camera angles |
| `zoom` | Zoom level, bounded 0.06 to 1 on import | Always |
| `tle` | External TLE source | When external-only is set, or a non-keeptrack source is active and `isMaxData` is true |
| `external-only` | Use external TLEs only | Written alongside `tle` when external-only is set |

## Extended export only (gated by `isShowExtendedUrlParams`)

These import normally but are only written back when the extended flag is on. Several only ever serialize their disabled state, so the enabled default stays implicit.

| Param | Notes |
| --- | --- |
| `vimpel` | Import sets JSC catalog on. Export only writes `vimpel=false` |
| `sensors` | Export only writes `sensors=false` |
| `launchSites` | Export only writes `launchSites=false` |
| `bottomMenu` | Export only writes `bottomMenu=false` |
| `camDistBuffer` | Written when a satellite is selected |
| `color` | Color scheme shorthand. Skipped when the scheme is the default `CelestrakColorScheme` |
| `ecf` | ECF orbit count. Written only when the count is not 1 |

## Import only (never written back)

`plugins` (enables the listed plugins and disables the rest), `canvas`, `planets`, `gp` (primary TLE source), `earth`, `sun`, `dots`, `msbai` (enables the OrbitGuard menu and maneuver plugins), `intldes` (international designator to selection), `misl` (missile simulation), `lat` / `lon`, plus the dev-only `debugFailTexture`, `debugFailStatus`, and `debugFailCount`.

## Color scheme shorthands (`color=`)

`type`, `celestrak`, `country`, `rcs`, `mission`, `confidence`, `orbitalplane`, `spatial`, `sunlight`, `gpage`, `source`, `velocity`, `starlink`, `smallsat`

## Mechanics worth knowing

- **Length cap.** `MAX_URL_LENGTH_` is 2000. When exceeded, heavy params are dropped in order: `tle`, then `external-only`, then `search`, then `limitSats`. If the URL is still too long, the update is skipped entirely to avoid 431 errors.
- **Throttle.** `updateURL` no-ops if called within 250 ms of the last write.
- **Kill switches.** `settingsManager.isDisableUrlBar` blocks all exports. A missing `searchManager` aborts the export early.

## Notable asymmetries

- `intldes` imports but only `sat` exports, so a selection round-trips as an SCC number.
- `gp` and `tle` are separate import paths, but only `tle` is exported.
- `vimpel`, `sensors`, `launchSites`, and `bottomMenu` only serialize their disabled state.
